File: others/problems/DP/purchase/Purchase.java
import java.util.Collections;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.stream.Collectors;

/*
 * Problem: Given multiple product discount plans,
 *   for e.g. sell products A, B, C together at price X
 * For a wish list of product type [M, N, ...] find the cheapest purchase plan
 *
 * For example, with the below plans:
 */

public class Purchase {
    private static final Map<String, Integer> TEST_PLAN1 =
        Map.of("AB",  100,
               "BCD", 150,
               "AD",  125,
               "CD",  135);

    /*
     * Below are the cheapest purchase plan:
     * lowest cost of "BAD"  # ["AB", "AD"]  ==> 225
     * lowest cost of "BAC"  # ["AB", "CD"]  ==> 235
     * lowest cost of "BCD"  # ["BCD"] ==> 150
     *
     * Here is another example. We use 0-9A-Z to enumerate products:
     */

    private static final Map<String, Integer> TEST_PLAN2 =
        Map.of("816309", 11,
               "7824", 28,
               "487i620", 47,
               "649", 57,
               "407396812", 57,
               "986750123", 64,
               "9642", 86,
               "16480579", 107,
               "9648350", 111,
               "8937514", 128);

    /*
     * the optimal purchase for product list "704938521" is
     * ["986750123", "7824"] ==> 28 + 64 = 92
     */

    private final static Set<String> EMPTY = Collections.EMPTY_SET;

    private static Set<String> empty() { return new HashSet<>(); }

    /*
     * Dynamic Programming solution
     * accepts discount plan, returns the DP table {cost: [products]}
     *
     * TODO: support multiple solution at the same cost, refer to the Haskell program
     */
    public static Map<Integer, Set<String>> dp(Map<String, Integer> plans) {
        Map<Integer, Set<String>> tab = new HashMap<>() {{
                put(0, EMPTY);
            }};
        for (String pkg : plans.keySet()) {
            for (Integer cost : new HashSet<>(tab.keySet())) {
                Set<String> pkgs = new HashSet<>(tab.get(cost));
                cost += plans.get(pkg);
                if (!(tab.containsKey(cost) &&
                      union(pkgs).containsAll(strToSet(pkg)))) {
                    pkgs.add(pkg);
                    tab.put(cost, pkgs);
                }
            }
        }
        return tab;
    }

    public static Set<String> lowest(Set<Character> prods,
                                     Map<Integer, Set<String>> tab) {
        return tab.keySet().stream().sorted()
            .map(p -> tab.get(p))
            .filter(s -> union(s).containsAll(prods))
            .findFirst().orElse(EMPTY);
    }

    private static Set<Character> union(Set<String> set) {
        Set<Character> s = new HashSet<>();
        for (String str : set) {
            s.addAll(strToSet(str));
        }
        return s;
    }

    private static Set<Character> strToSet(String s) {
        return s.chars().mapToObj(c -> (char) c)
            .collect(Collectors.toSet());
    }

    /*
     * DFS based Brute force solution for verification purpose
     */
    public static Set<String> findLowest(Map<String, Integer> plan,
                                         Set<Character> wish) {
        Set<String> best = empty();
        dfs(plan, wish, empty(), best, Integer.MAX_VALUE);
        return best;
    }

    private static int dfs(Map<String, Integer> plan, Set<Character> wish,
                           Set<String> res,
                           Set<String> best, int minSofar) {
        if (wish.isEmpty()) {
            int cost = costOf(res, plan);
            if (cost < minSofar) {
                //System.out.format("find a better %s at %d\n", res.toString(), cost);
                best.clear();
                best.addAll(res);
                minSofar = cost;
            }
        } else {
            char p = wish.iterator().next();
            for (String pkg : plan.keySet()) {
                if (pkg.indexOf(p) != -1) {
                    res.add(pkg);
                    minSofar = dfs(plan, diff(wish, strToSet(pkg)), res, best, minSofar);
                    res.remove(pkg);
                }
            }
        }
        return minSofar;
    }

    private static Set<Character> diff(Set<Character> a, Set<Character> b) {
        Set<Character> c = new HashSet<>(a);
        c.removeAll(b);
        return c;
    }

    private static int costOf(Set<String> pkg, Map<String, Integer> plan) {
        int c = 0;
        for (String p : pkg) {
            c += plan.get(p);
        }
        return c;
    }

    /*
     * None recursive Brute force solution
     */
    public static Set<String> findMin(Map<String, Integer> plan,
                                      Set<Character> wish) {
        Set<Set<String>> res = new HashSet<>();
        res.add(EMPTY);
        for (Character p : wish) {
            Set<String> pkgs = empty();
            for (String pkg : plan.keySet()) {
                if (pkg.indexOf(p) != -1) {
                    pkgs.add(pkg);
                }
            }
            res = product(res, pkgs);
            if (res.isEmpty()) {
                return EMPTY;
            }
        }
        Set<String> best = EMPTY;
        int lowest = Integer.MAX_VALUE;
        for (Set<String> p : res) {
            int cost = costOf(p, plan);
            if (cost < lowest) {
                best = p;
                lowest = cost;
            }
        }
        return best;
    }

    // Cartesian product
    private static <T> Set<Set<T>> product(Set<Set<T>> sets, Set<T> set) {
        Set<Set<T>> res = new HashSet<>();
        for (Set<T> a : sets) {
            for (T b : set) {
                Set<T> c = new HashSet<>(a);
                c.add(b);   // c = a `union` b
                res.add(c);
            }
        }
        return res;
    }

    /*
     * Verification
     */
    private static void verify(String wish, Map<Integer, Set<String>> tab,
                               Map<String, Integer> plan) {
        Set<Character> wishSet = strToSet(wish);
        System.out.format("dfs for %s ==> %s\n", wish,
                          findLowest(plan, wishSet).toString());
        System.out.format("dp  for %s ==> %s\n", wish,
                          lowest(wishSet, tab).toString());
        System.out.format("bfs for %s ==> %s\n", wish,
                          findMin(plan, wishSet).toString());
    }

    public static void main(String[] args) {
        Map<String, Integer> plan = TEST_PLAN1;
        Map<Integer, Set<String>> tab = dp(plan);
        verify("BAD", tab, plan);
        verify("BAC", tab, plan);
        verify("BCD", tab, plan);
        plan = TEST_PLAN2;
        tab = dp(plan);
        verify("704938521", tab, plan);
    }
}

File: others/problems/DP/purchase/Purchase.hs
module Purchase where

import Data.Map as Map
import Data.Set as Set

-- Problem: Given multiple product discount plans,
--   for e.g. sell products A, B, C together at price X
-- For a wish list of product type [M, N, ...] find the cheapest purchase plan
--
-- For example, with the below plans:

plan1 = Map.fromList [("AB",  100)
                     ,("BCD", 150)
                     ,("AD",  125)
                     ,("CD",  135)]

-- Below are the cheapest purchase plan:
-- lowest cost of "BAD"  # ["AB", "AD"]  ==> 225
-- lowest cost of "BAC"  # ["AB", "CD"]  ==> 235
-- lowest cost of "BCD"  # ["BCD"] ==> 150

-- Here is another example. We use 0-9A-Z to enumerate products:

plan2 = Map.fromList [("816309", 11)
                     ,("7824", 28)
                     ,("487i620", 47)
                     ,("649", 57)
                     ,("407396812", 57)
                     ,("986750123", 64)
                     ,("9642", 86)
                     ,("16480579", 107)
                     ,("9648350", 111)
                     ,("8937514", 128)]

-- the optimal purchase for product list "704938521" is
-- ["986750123", "7824"] ==> 28 + 64 = 92

type Plan = Map String Integer  -- Map package price
type Packages = Set String

-- Different Packages may cost same, hence we use Set Packages.

type DpTab = Map Integer (Set Packages)  -- Map cost (Set Packages)

-- Dynamic Programming solution
--  Build the DP table
dpTable :: Plan -> DpTab
dpTable = foldrWithKey build (Map.singleton 0 Set.empty)

build :: String -> Integer -> DpTab -> DpTab
build pkg price tab = foldrWithKey expand tab tab where
  expand cost pkgss tab = if Set.null pkgss' then tab
                          else insertWith Set.union cost' pkgss' tab where
    cost' = cost + price
    pkgss' = pkg `exInsert` pkgss

-- Exclusive Insert, for each set of packages combination, only insert the
-- packages when it can't be covered
exInsert :: String -> Set Packages -> Set Packages
exInsert pkg pkgss
  | Set.null pkgss = Set.singleton (Set.singleton pkg)
  | otherwise = Set.map (Set.insert pkg) $ Set.filter (not . exists pkg) pkgss

-- if a package can be fully covered by a set of packages combination
exists :: String -> Packages -> Bool
exists pkg pkgs = (Set.fromList pkg) `isSubsetOf` (Set.unions $ Set.map Set.fromList pkgs)

-- Lookup the DP table for the lowest cost purchase plans (multiple results)
lowest :: String -> DpTab -> Maybe (Integer, Set Packages)
lowest wish = filterPkg . Map.lookupMin . (Map.filter (any (exists wish))) where
  filterPkg Nothing = Nothing
  filterPkg (Just (cost, pkgss)) = Just (cost, Set.filter (exists wish) pkgss)

-- examples
tab1 = dpTable plan1
test1 = Prelude.map (\s -> lowest s tab1) ["BAD", "BAC", "BCD"]

tab2 = dpTable plan2
test2 = lowest "704938521" tab2
